File: apps/ai-workers/app/core/event_logger.py
from __future__ import annotations

import json
import uuid
from typing import Any

from app.core.redis_client import client


def ensure_trace_id(trace_id: str | None = None) -> str:
    return trace_id or str(uuid.uuid4())


def log_worker_event(
    *,
    trace_id: str | None,
    service: str,
    stage: str,
    event_type: str,
    level: str = "info",
    user_id: str | None = None,
    interview_id: str | None = None,
    file_id: str | None = None,
    payload: dict[str, Any] | None = None,
) -> str:
    safe_trace_id = ensure_trace_id(trace_id)
    client.publish(
        "event:log",
        json.dumps(
            {
                "traceId": safe_trace_id,
                "service": service,
                "stage": stage,
                "eventType": event_type,
                "level": level,
                "userId": user_id,
                "interviewId": interview_id,
                "fileId": file_id,
                "payload": payload or {},
            }
        ),
    )
    return safe_trace_id

File: apps/ws-backend/src/utils/eventLogger.ts
import { randomUUID } from "crypto";
import { redisClient } from "../config/redis.config.js";

export type EventLogLevel = "DEBUG" | "INFO" | "WARN" | "ERROR";

export interface EventLogEnvelope {
  traceId?: string | null;
  service?: string;
  stage: string;
  eventType: string;
  level?: EventLogLevel;
  userId?: string | null;
  interviewId?: string | null;
  fileId?: string | null;
  payload?: Record<string, unknown> | null;
}

export async function logEvent(event: EventLogEnvelope): Promise<string> {
  const traceId = event.traceId || randomUUID();
  const envelope = {
    traceId,
    service: event.service ?? "ws-backend",
    stage: event.stage,
    eventType: event.eventType,
    level: event.level ?? "INFO",
    userId: event.userId ?? null,
    interviewId: event.interviewId ?? null,
    fileId: event.fileId ?? null,
    payload: event.payload ?? null,
    createdAt: new Date().toISOString(),
  };

  try {
    await redisClient.publish("event:log", JSON.stringify(envelope));
  } catch (error) {
    console.error("[eventLogger] publish failed:", error);
  }

  return traceId;
}
